### DISCUSSION TITLE
ArrayBuffer 增强

### DISCUSSION THREAD

**Discussion by septs** [JSCIG/es-discuss#18] [Ideas]
```typescript
// 等效 `.compare(a, b) === 0` 
ArrayBuffer.equals(a: ArrayBuffer, b: ArrayBuffer): boolean

ArrayBuffer.compare(a: ArrayBuffer, b: ArrayBuffer): -1 | 0 | 1

ArrayBuffer.from(input: string, type: "base64" | "hex"): ArrayBuffer

ArrayBuffer.prototype.toString(type?: "base64" | "hex" = "base64"): string

// 等效 `.toString("base64")` 
ArrayBuffer.prototype.toJSON(): string
```

1. 避免总是引入和重复实现 Base 64 / Hex 编解码
2. 已知 `atob` / `btoa` 不在 ES 规范中
3. 提供 **比较相等** 和 **排序** 的方法

**Comment by codehz** [comment]
不如弄个BinaryEncoder/BinaryDecoder，其中可以为hex或者base64，以及其他扩展格式

**Comment by septs** [comment]
> 不如弄个BinaryEncoder/BinaryDecoder，其中可以为hex或者base64，以及其他扩展格式

就 https://mdn.io/TextEncoder / https://mdn.io/TextDecoder 来说并不具备任何扩展意义

当然如果 BinaryEncoder / BinaryDecoder 从设计上具备扩展意义
那么相对的 TextEncoder / TextDecoder 也需要相应的设计

建构在 ArrayBuffer 基础上，也是可以方便未来将其推广到 TypedArray 之上

**Comment by septs** [comment]
`toJSON()` 想法取消，break change

**Comment by hax** [comment]
`compare` 的 -1 和 +1 表示啥？

**Comment by septs** [comment]
> `compare` 的 -1 和 +1 表示啥？

我参考的是 Node.js 的 Buffer.compare 设计

https://nodejs.org/api/buffer.html#buffer_buf_compare_target_targetstart_targetend_sourcestart_sourceend

**Comment by hax** [comment]
我就是不太清楚对 buffer 进行 sort 是有什么使用场景。也许经常搞 node 的同学知道。

**Comment by hax** [comment]
base64现在已经有提案了，不过是在Uint8Array上。（其实顶楼建议的这些方法都应该在TypedArray而不是在ArrayBuffer上。）
